lib/plugindata/cas.go

```diff
@@ -0,0 +1,129 @@
+package plugindata
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/teleport-plugins/lib/backoff"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+)
+
+const (
+	// backoffBase is an initial (minimum) backoff value.
+	backoffBase = time.Millisecond
+	// backoffMax is a backoff threshold
+	backoffMax = time.Second
+)
+
+// Client represents an interface to Teleport API client
+type Client interface {
+	GetPluginData(context.Context, types.PluginDataFilter) ([]types.PluginData, error)
+	UpdatePluginData(context.Context, types.PluginDataUpdateParams) error
+}
+
+// EncodeFn represents a function type of plugin data serializer
+type EncodeFn = func(interface{}) map[string]string
+
+// DecodeFn represents a function type of plugin data deserializer
+type DecodeFn = func(map[string]string) interface{}
+
+// Cb represents a type for Plugin Data modification fn
+type Cb = func(interface{}) (interface{}, bool)
+
+// CAS represents modifier struct
+type CAS struct {
+	client      Client
+	name        string
+	kind        string
+	backoffBase time.Duration
+	backoffMax  time.Duration
+	encode      EncodeFn
+	decode      DecodeFn
+}
+
+// NewCAS returns modifier struct
+func NewCAS(client Client, name, kind string, encode EncodeFn, decode DecodeFn) *CAS {
+	return &CAS{
+		client,
+		name,
+		kind,
+		backoffBase,
+		backoffMax,
+		encode,
+		decode,
+	}
+}
+
+// Modify performs a compare-and-swap update of a resource plugin data.
+//
+// fn callback function receives current plugin data value and returns modified value and
+// a flag, which indicates that input value was valid and up to date, and the resulting value
+// must be saved.
+//
+// Please note that fn might be called several times due to CAS backoff, hence, you must be careful
+// with things like I/O ops and channels.
+//
+// Method returns modified plugin data on success, nil when data was not saved, error otherwise.
+//
+// container argument must contain the reference to an empty struct implementing PluginData interface.
+func (c *CAS) Modify(ctx context.Context, resource string, cb Cb) (interface{}, error) {
```

Comment:
Can we split this into `Modify` and `Add`, where `Modify` returns a `trace.NotFound` if `oldData == nil` and `Add` returns a `trace.AlreadyExists` if `oldData != nil`? Then callers can check relevant errors instead of checking `pluginData == nil`.

I think this would be more straightforward and easy to use, which is important given this is a new library.

Reply:
Compare-and-swap supposed to be a single atomic instruction. I possibly get your idea wrong, but splitting this into two methods would break the atomicity. Just in case, this method is not new. It is used in every plugin like this, I just tried to DRY it into a separate module.

Reply:
I agree that we should use compare-and-swap, but we can still move the functionality from the callback to the function by splitting it into two functions. Here's what I mean:

```
func (c *CAS) Modify(ctx context.Context, resource string, modify func(interface{}) interface{}) (interface{}, error) {
	backoff := backoff.NewDecorr(c.backoffBase, c.backoffMax, clockwork.NewRealClock())
	for {
		oldData, err := c.getPluginData(ctx, resource)
		if err != nil {
			return nil, trace.Wrap(err)
		}

		newData := modify(oldData)

		err = trace.Wrap(c.updatePluginData(ctx, resource, newData, oldData))
		if err == nil {
			return newData, nil
		}
		if !trace.IsCompareFailed(err) {
			return nil, trace.Wrap(err)
		}
		if err := backoff.Do(ctx); err != nil {
			return nil, trace.Wrap(err)
		}
	}
}

func (c *CAS) Create(ctx context.Context, resource string, pd interface{}) (interface{}, error) {
	// We don't need to do backoff, because if compare fails
	// that means the plugin data already exists. Abort creation.
	err := trace.Wrap(c.updatePluginData(ctx, resource, pd, nil))
	if err != nil {
		return nil, trace.Wrap(err)
	}
	return pd, nil
}
```

Now we can simplify our calls to Modify like this:

```
// postMessages posts initial Teams messages
func (a *App) postMessages(ctx context.Context, recipients []string, id string, data pd.AccessRequestData) error { 
	...
	// Let's update sent messages data
	_, err = a.pd.Modify(ctx, id, func(existing interface{}) (interface{}) {
		pluginData, ok := existing.(PluginData)
		if !ok {
			// maybe return error?
			return nil
		}
		pluginData.TeamsData = teamsData
		return pluginData
	})
	return trace.Wrap(err)
}
```

And we can simplify Creation like this:

```
func (a *App) onPendingRequest(ctx context.Context, req types.AccessRequest) error {
	...
	pd, err := a.pd.Modify(ctx, id, PluginData{AccessRequestData: data})
	if err != nil {
		return trace.Wrap(err)
	}
```

Reply:
Ah, now I see. I've already tried to achieve the same goal with `mode` argument of the method. There are actually three situations: "Create", "Update" and "Upsert". 

Let me finish my refactoring, I get back to you and we discuss an updated version.

Reply:
I guess, I can close this conversation